### ISSUE TITLE
Delayed registration?

### ISSUE THREAD

**Issue by hadley** [RConsortium/OOP-WG#5]
Should we explicitly consider delayed registration in the requirements? This is needed if you want to provide a method for a generic (or class) that is available in a suggested (not imported) package. 

[R-exsts](https://cran.rstudio.com/doc/manuals/r-devel/R-exts.html#Registering-S3-methods) has:

> As from R 3.6.0 one can also use S3method() directives to perform delayed registration. With
> 
> if(getRversion() >= "3.6.0") {
>    S3method(pkg::gen, cls)
> }

**Comment by lawremi** [comment]
Yes, this would be useful. Taking the object-oriented approach, we could have a proxy object representing the generic/class. It would react to the loading of the actual definition. Perhaps constructed like:
```
MyClass <- make_class(pkg::SuperClass, ...)
```
Just as with `S3method()`, it would see the namespace qualifier and default to laziness.

**Comment by hadley** [comment]
I just remembered there's another case where you need condition method registration — when you're releasing a package that works with multiple versions of a dependency. 

For example, dplyr 1.0.0 added a new generic `dplyr_reconstruct()`. If your pacakge extends dplyr, you'd like to be able to release a new version that works with both dplyr 0.8.5 and dplyr 1.0.0, before dplyr 1.0.0 is released. That means you want to register the method only if dplyr 1.0.0 or later is available when your package is loaded.

**Comment by lawremi** [comment]
That may be the only practical solution; however, it would obviously be ideal if you could test against a branch of the dependency and then coordinate the release.

Anyway, since registration would happen in regular R code, it could operate conditionally, although if we're doing any caching it could get messy.

**Comment by hadley** [comment]
That also forces the users of the package to update to the latest version of its dependency; that's not always great if it's mostly a data-analyst facing package.

**Comment by mmaechler** [comment]
> I just remembered there's another case where you need condition method registration — when you're releasing a package that works with multiple versions of a dependency.
> 
> For example, dplyr 1.0.0 added a new generic `dplyr_reconstruct()`. If your pacakge extends dplyr, you'd like to be able to release a new version that works with both dplyr 0.8.5 and dplyr 1.0.0, before dplyr 1.0.0 is released. That means you want to register the method only if dplyr 1.0.0 or later is available when your package is loaded.

I vaguely remember this almost works now already in NAMESPACE, as it allows conditionals;  however that of course maybe at the wrong point in time: The (non-)registration should happen at package load time, not at build time.
